feature_details.md

```diff
@@ -0,0 +1,413 @@
+
+An important assumption in the design of this feature is that it's reasonable to limit customization options for cross-origin transitions to make it easier for UAs to enforce the constraints above. But the API/implementation should strive to not unnecessarily limit capabilities for same-origin transitions due to cross-origin constraints.
+
+## Glossary
+| Term | Description |
+| ------------- | ------------- |
+| Previous Document  | The Document the user is viewing when a navigation is initated.  |
+| New Document  | The Document which will be current in the session history when a navigation is committed. |
+| Root transition | An animation to transition the content for root element of both Documents. |
+| Exit transition | An animation which specifies how an element in the previous Document exits the screen. |
+| Enter transition | An animation which specifies how an element in the new Document enters the screen. |
+| Snapshot | A pixel copy for a DOM subtree. |
+
+## Sample Code
+This section explains how the feature can be used to add transitions to MPAs. There are a set of use-cases and edge cases to consider :
+
+### Root Transition
+Let's start with a root transition, which adds an animation to the root element of previous and new Documents. Below is a video of the transition we want to create :
+
+[![Video Link for Root Element Transition](https://img.youtube.com/vi/mCvvWMIGItY/0.jpg)](https://www.youtube.com/watch?v=mCvvWMIGItY)
+
+The entry point for this API is a new `document.documentTransition` object. The previous Document provides a list of elements which will be animated during the transition. Since this is root transition, this list only includes the `html` element. The result of calling `documentTransition.setSameOriginTransitionElements` as illustrated below is that the browser caches a copy of the pixels (snapshot) for the `html` element when a navigation to `foo.com/a.html` is initiated.
+
+```
+// The list of elements is keyed on url. When a user navigates to
+// a url, the corresponding entry in this dictionary is used to
+// decide which elements should be cached.
+document.documentTransition.setSameOriginTransitionElements(
+  {
+   “foo.com/a.html” : [{element : document.html}],
+  });
+```
+
+Now we need to add some code on the new Document. When the new Document loads, script registers a listener for a new event type called `handleTransition`. The event is dispatched if a transition was initiated by the previous Document. The exact Document lifecycle stage before which this event should be registered is TBD.
+
+```
+document.documentTransition.addEventListener(
+    “handleTransition”, e => {
```

Comment:
One thing we need to worry about is the event firing before the event handler is registered. For example, if we fire the event after the above-the-fold content has been parsed and presented, and the event handler is at the end of the page that hasn't been parsed yet, then we will miss the event.

Nothing to change in the doc here, but just something to keep in mind.

Reply:
Noted. With respect to timing for this event, I assumed the trickier problem is to identify the stage in the navigation lifecycle when it is appropriate to dispatch this event. Some notes from digging right now, it needs to be between step 5/6 in the history traversal algorithm : https://html.spec.whatwg.org/multipage/browsing-the-web.html#history-traversal. I expect this needs to be fired between the document.readyState changing from loading to interactive : https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState

I didn't realize the position in the page where tha handler is registered could also be a difficult aspect for the developer to reason about. Thanks for pointing this out!

Reply:
Added a TODO in API proposal to make a note of this.